### PR TITLE
p2p: implement interface for p2p.Channel without channels

### PR DESCRIPTION
**Pull request by tychoish**
This provides an alternate implementation for `p2p.Channel`s, for use
in reactors. If this makes sense to people, the next step is to make
each field in `p2p.Channel` private and then replace each with these
calls.

### CODE REVIEW

File: internal/p2p/channel.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+// concept that can mark good and bad behavior and contributes to peer scoring.
+// It should possibly also allow reactors to request explicit actions, e.g.
+// disconnection or banning, in addition to doing this based on aggregates.
+type PeerError struct {
+	NodeID types.NodeID
+	Err    error
+}
+
+func (pe PeerError) Resolve() error { return fmt.Errorf("peer=%q: %w", pe.NodeID, pe.Err) }
+func (pe PeerError) Error() string  { return pe.Resolve().Error() }
```
Comment (creachadair) [inline internal/p2p/channel.go:59]: Maybe consider implementing Unwrap so it will play nicely with `errors.Is` and `errors.As`?

```suggestion
func (pe PeerError) Error() string  { return pe.Resolve().Error() }
func (pe PeerError) Unwrap() error  { return pe.Err }
```

File: internal/p2p/channel.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+// FIXME: This should probably be replaced by a more general PeerBehavior
+// concept that can mark good and bad behavior and contributes to peer scoring.
+// It should possibly also allow reactors to request explicit actions, e.g.
+// disconnection or banning, in addition to doing this based on aggregates.
+type PeerError struct {
+	NodeID types.NodeID
+	Err    error
+}
+
+func (pe PeerError) Resolve() error { return fmt.Errorf("peer=%q: %w", pe.NodeID, pe.Err) }
```
Comment (creachadair) [inline internal/p2p/channel.go:58]: I'm not sure why this method is here—it doesn't seem to be used apart from in `Error` itself. Do we need the extra layer?

Already a `PeerError` can be returned as an `error` even without this.

File: internal/p2p/channel.go [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+func (pe PeerError) Error() string  { return pe.Resolve().Error() }
+
+// Channel is a bidirectional channel to exchange Protobuf messages with peers,
+// wrapped in Envelope to specify routing info (i.e. sender/receiver).
```
Comment (creachadair) [inline internal/p2p/channel.go:62]: Minor nit: Avert confusion about whether the channel or the message is wrapped:
```suggestion
// Channel is a bidirectional channel to exchange Protobuf messages with peers.
// Each message is wrapped in an Envelope to specify its sender and receiver.
```

File: internal/p2p/channel.go [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+type Channel struct {
+	ID    ChannelID
+	In    <-chan Envelope  // inbound messages (peers to reactors)
+	Out   chan<- Envelope  // outbound messages (reactors to peers)
+	Error chan<- PeerError // peer error reporting
+
+	messageType proto.Message // the channel's message type, used for unmarshaling
```
Comment (creachadair) [inline internal/p2p/channel.go:69]: If we also exported this field, we wouldn't need a separate constructor. 

There doesn't seem much harm in exporting it—if the caller can be trusted with the channels, exporting the message type seems fine. It doesn't have any subtle semantics.

File: internal/p2p/channel.go [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+	return &Channel{
+		ID:          id,
+		messageType: messageType,
+		In:          inCh,
+		Out:         outCh,
+		Error:       errCh,
+	}
+}
+
+func (ch *Channel) Send(ctx context.Context, envelope Envelope) error {
```
Comment (creachadair) [inline internal/p2p/channel.go:90]:
```suggestion
// Send blocks until the envelope has been sent, or until ctx ends.
// An error only occurs if the context ends before the send completes.
func (ch *Channel) Send(ctx context.Context, envelope Envelope) error {
```

File: internal/p2p/channel.go [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case ch.Out <- envelope:
+		return nil
+	}
+}
+
+func (ch *Channel) SendError(ctx context.Context, pe PeerError) error {
```
Comment (creachadair) [inline internal/p2p/channel.go:99]:
```suggestion
// SendError blocks until the given error has been sent, or ctx ends.
// An error only occurs if the context ends before the send completes.
func (ch *Channel) SendError(ctx context.Context, pe PeerError) error {
```

File: internal/p2p/channel.go [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case ch.Error <- pe:
+		return nil
+	}
+}
+
+func (ch *Channel) Receive(ctx context.Context) *ChannelIterator {
```
Comment (creachadair) [inline internal/p2p/channel.go:108]:
```suggestion
// Receive returns a new unbuffered iterator to receive messages from ch.
// The iterator runs until ctx ends.
func (ch *Channel) Receive(ctx context.Context) *ChannelIterator {
```

File: internal/p2p/channel.go [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+// Envelope returns the current Envelope object held by the
+// iterator. When the last call to Next returned true, Envelope will
+// return a non-nil object. If Next returned false then Envelope is
+// always nil.
+func (iter *ChannelIterator) Envelope() *Envelope { return iter.current }
+
+// MergedChannelIterator produces a Receive iterator from multiple
+// channels to support the reactor use case of needing to consume
+// messages from multiple sources in a single reactor without needing
+// the reactor to manage the concurrency for these handlers.
```
Comment (creachadair) [inline internal/p2p/channel.go:180]: Wording suggestion, to separate what it does from why:

```suggestion
// MergedChannelIterator produces an iterator that merges the
// messages from the given channels in arbitrary order.
//
// This allows the caller to consume messages from multiple channels
// without needing to manage the concurrency separately.
```

File: internal/p2p/channel.go [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+func MergedChannelIterator(ctx context.Context, chs ...*Channel) *ChannelIterator {
+	iter := &ChannelIterator{
+		pipe: make(chan Envelope), // unbuffered
+	}
+	wg := &sync.WaitGroup{}
```
Comment (creachadair) [inline internal/p2p/channel.go:185]: I suggest:
```suggestion
	var wg sync.WaitGroup
```

Or if you prefer `wg := new(sync.WaitGroup)` (since this type will never have fields we can initialize)

File: internal/p2p/channel.go [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+	wait := make(chan struct{})
+	go func() { defer close(wait); wg.Wait() }()
+
+	go func() {
+		defer close(iter.pipe)
+		// we could return early if the context is canceled,
+		// but this is safer because it means the pipe stays
+		// open until all of the ch worker threads end, which
+		// should happen very quickly.
+		<-wait
```
Comment (creachadair) [inline internal/p2p/channel.go:196]: Nit: naming suggestion.
```suggestion
	done := make(chan struct{})
	go func() { defer close(done); wg.Wait() }()

	go func() {
		defer close(iter.pipe)
		// we could return early if the context is canceled,
		// but this is safer because it means the pipe stays
		// open until all of the ch worker threads end, which
		// should happen very quickly.
		<-done
```

File: internal/p2p/channel.go [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+	}()
+
+	for idx := range chs {
+		wg.Add(1)
+		go func(ch *Channel) {
+			defer wg.Done()
+			iteratorWorker(ctx, ch, iter.pipe)
+		}(chs[idx])
+	}
```
Comment (creachadair) [inline internal/p2p/channel.go:205]:
```suggestion
	for _, ch := range chs {
		wg.Add(1)
		go func(ch *Channel) {
			defer wg.Done()
			iteratorWorker(ctx, ch, iter.pipe)
		}(ch)
	}
```

File: internal/p2p/channel_test.go [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+					require.Fail(t, "iterator should not advance")
+				}
+				require.Nil(t, iter.Envelope())
+			},
+		},
+		{
+			Name: "ReceiveWithData",
+			Case: func(ctx context.Context, t *testing.T) {
+				ins, ch := testChannel(1)
+				ins.In <- Envelope{From: "kip", To: "merlin"}
```
Comment (creachadair) [inline internal/p2p/channel_test.go:112]: 😻

File: internal/p2p/channel_test.go [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+	for idx := range testCases {
+		t.Run(testCases[idx].Name, func(t *testing.T) {
+			t.Cleanup(leaktest.Check(t))
+
+			ctx, cancel := context.WithCancel(bctx)
+			defer cancel()
+
+			testCases[idx].Case(ctx, t)
+		})
+	}
```
Comment (creachadair) [inline internal/p2p/channel_test.go:182]:
```suggestion
	for _, tc := range testCases {
		t.Run(tc.Name, func(t *testing.T) {
			t.Cleanup(leaktest.Check(t))

			ctx, cancel := context.WithCancel(bctx)
			defer cancel()

			tc.Case(ctx, t)
		})
	}
```

**Review by creachadair — APPROVED** [review]
This seems quite good, I have a few suggestions but nothing major. I especially appreciate the test coverage here.

File: internal/p2p/channel_test.go (hunk 13 above)
Comment (creachadair) [reply internal/p2p/channel_test.go:182]: Sure, but that just means it's confusing now, and we'd have to do other things to make these tests parallelizable anyway. That said, I don't object, it just seemed unnecessary 🙂

File: internal/p2p/channel_test.go [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+
+				res := iter.Envelope()
+				require.EqualValues(t, "kip", res.From)
+				require.EqualValues(t, "merlin", res.To)
+
+				cctx, ccancel := context.WithCancel(ctx)
+				ccancel()
+
+				require.False(t, iter.Next(cctx))
+				require.Nil(t, iter.Envelope())
```
Comment (williambanfield) [inline internal/p2p/channel_test.go:168]: Really appreciate these tests. I like the semantics of having `Next` nil out the value when the ctx is canceled.

File: internal/p2p/channel_test.go [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+				require.NotNil(t, iter)
+
+				cctx, ccancel := context.WithCancel(ctx)
+				ccancel()
+				require.False(t, iter.Next(cctx))
+				require.Nil(t, iter.Envelope())
+			},
+		},
+		{
+			Name: "IteratorCanceledAfterFirstUseBecomesNil",
```
Comment (williambanfield) [inline internal/p2p/channel_test.go:150]: Should we add a test for `Envelope before next is nil`?

File: internal/p2p/channel.go [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+				return
+			case pipe <- envelope:
+			}
+		}
+	}
+}
+
+// Next returns true when the Envelope value has advanced, and false
+// when the context is canceled or iteration should stop, then Next
+// will return true. In general, use Next, as in:
```
Comment (williambanfield) [inline internal/p2p/channel.go:147]:
```suggestion
// Next returns true when the Envelope value has advanced, and false
// when the context is canceled or iteration should stop. If an iterator has returned false,
// it will never return true again.
// in general, use Next, as in:
```

**Review by williambanfield — APPROVED** [review]
